plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/mutate/MutateProcessor.java

```diff
@@ -0,0 +1,312 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.ingest.processor.mutate;
+
+import org.elasticsearch.ingest.Data;
+import org.elasticsearch.ingest.processor.Processor;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+public final class MutateProcessor implements Processor {
+
+    public static final String TYPE = "mutate";
+
+    private final Map<String, Object> update;
+    private final Map<String, String> rename;
+    private final Map<String, String> convert;
+    private final Map<String, String> split;
+    private final Map<String, List<String>> gsub;
+    private final Map<String, String> join;
+    private final List<String> remove;
+    private final List<String> strip;
+    private final List<String> uppercase;
+    private final List<String> lowercase;
+
+    public MutateProcessor(Map<String, Object> update,
+                           Map<String, String> rename,
+                           Map<String, String> convert,
+                           Map<String, String> split,
+                           Map<String, List<String>> gsub,
+                           Map<String, String> join,
+                           List<String> remove,
+                           List<String> strip,
+                           List<String> uppercase,
+                           List<String> lowercase) {
+        this.update = update;
+        this.rename = rename;
+        this.convert = convert;
+        this.split = split;
+        this.gsub = gsub;
+        this.join = join;
+        this.remove = remove;
+        this.strip = strip;
+        this.uppercase = uppercase;
+        this.lowercase = lowercase;
+    }
+
+    @Override
+    public void execute(Data data) {
+        doUpdate(data);
+        doRename(data);
+        doConvert(data);
+        doSplit(data);
+        doGsub(data);
+        doJoin(data);
+        doRemove(data);
+        doStrip(data);
+        doUppercase(data);
+        doLowercase(data);
+    }
+
+    private void doUpdate(Data data) {
+        if (update == null) { return; }
```

Comment:
maybe move all these if checks to execute method? (I personally find that easier to read, when we get in do* method we know we're good to go) Like this:

```java
if (update != null) {
   doUpdate(data);
}
// next if
```